app/controllers/campaign_pages_controller.rb

```diff
@@ -43,6 +55,42 @@ def create
   end
 
   def show
-    @page = CampaignPage.find params[:id]
+    @campaign_page = CampaignPage.find params[:id]
+    if @campaign_page.active == false
+      redirect_to :campaign_pages, notice: "The page you wanted to view has been deactivated."
+    end  
   end
+
+  def edit
+    @campaign_page = CampaignPage.find params[:id]
+  end
+
+  def update
+    @campaign_page = CampaignPage.find params[:id]
+    @widgets = @campaign_page.campaign_pages_widgets
+
+    permitted_params = CampaignPageParameters.new(params).permit
+    permitted_params[:slug] = permitted_params[:title].parameterize
+    permitted_params[:campaign_pages_widgets_attributes] = []
+    params[:widgets].each do |widget_type_name, widget_data|
+      # widget type id is contained in a field called widget_type:
+      widget_type_id = widget_data.delete :widget_type
+      # This will break if there will be two widgets of the same type for the page. If we enable having two of the same widget type per page,
+      # page display order will need to be considered as well for fetching the correct id of the widget.
+      widget = @widgets.find_by(widget_type_id: widget_type_id)
+      permitted_params[:campaign_pages_widgets_attributes].push({
```

Comment:
Basically what's happening here is that we've got HTML placeholders for checkbox and textarea form elements that get dropped into the form (hidden) and then populated using Javascript if the user wants to add one to the form.

The result is that we've got to strip those values out of the form parameters before saving, or we wind up with the extras. There's unfortunately not a super clean way to do this no matter how we deal with it (we could strip the params before submitting the form, but that's not really better, it just bumps the lack of cleanliness to a different step in the process).

I'm thinking that we might want to pull it into the model, though, and build tests for it.